pygen/pygen_src/riscv_instr_stream.py

```diff
@@ -0,0 +1,230 @@
+     instruction, mix two instruction streams etc.
+    '''
+
+    def __init__(self):
+        self.instr_list = []
+        self.instr_cnt = 0
+        self.label = ""
+        # User can specify a small group of available registers to generate various hazard condition
+        self.avail_regs = []
+        # Some additional reserved registers that should not be used as rd register
+        # by this instruction stream
+        self.reserved_rd = []
+        self.hart = 0
+
+    def initialize_instr_list(self, instr_cnt):
+        self.instr_list = []
+        self.instr_cnt = instr_cnt
+        self.create_instr_instance()
+
+    def create_instr_instance(self):
+        for i in range(self.instr_cnt):
+            instr = riscv_instr()
+            self.instr_list.append(instr)
+
+    def insert_instr(self, instr, idx=-1):
+        """
+           Insert an instruction to the existing instruction stream at the given index
+           When index is -1, the instruction is injected at a random location
+        """
+        current_instr_cnt = len(self.instr_list)
+        if idx == -1:
+            idx = random.randint(0, current_instr_cnt - 1)
+            while self.instr_list[idx].atomic:
+                idx = idx + 1
+                if idx == (current_instr_cnt - 1):
+                    self.instr_list.append(instr)
+                    return
+        elif idx > current_instr_cnt or idx < 0:
+            # TO DO: print an error
+            print("Error!")
+        self.instr_list.insert(idx, instr)
+
+    def insert_instr_stream(self, new_instr, idx=-1, replace=0):
+        """
+            Insert an instruction to the existing instruction stream at the given index
+            When index is -1, the instruction is injected at a random location
+            When replace is 1, the original instruction at the inserted position will be replaced
+        """
+        current_instr_cnt = len(self.instr_list)
+
+        if current_instr_cnt == 0:
+            self.instr_list = new_instr
+            return
+
+        if idx == -1:
+            idx = random.randint(0, current_instr_cnt - 1)
+            # cares must be taken to avoid targeting
+            # an atomic instruction (while atomic, find a new idx)
+            for i in range(10):
+                if self.instr_list[idx].atomic:
+                    break
+                idx = random.randint(0, current_instr_cnt - 1)
+            if self.instr_list[idx].atomic:
+                for i in range(len(self.instr_list)):
+                    if not self.instr_list[i].atomic:
+                        idx = i
+                        break
+                if self.instr_list[idx].atomic:
+                    print("Error")  # TODO: Put appropriate print here
+        elif idx > current_instr_cnt or idx < 0:
+            # TODO: Print an error indicating that inserting an instruction
+            #       in this location is not possible
+            pass
```

Comment:
You can use logging.error()
and do sys.exit(1) after that line.

Reply:
all the logging related lines I have kept under TODO. Will update this now